service/dynamodb/expression/expression.go

```diff
@@ -0,0 +1,630 @@
+//
+//     queryInput := dynamodb.QueryInput{
+//       KeyConditionExpression:    factory.KeyCondition(),
+//       ProjectionExpression:      factory.Projection(),
+//       ExpressionAttributeNames:  factory.Names(),
+//       ExpressionAttributeValues: factory.Values(),
+//       TableName: aws.String("SomeTable"),
+//     }
+type Factory struct {
+	expressionMap map[expressionType]TreeBuilder
+}
+
+// TreeBuilder interface will be fulfilled by builder structs that
+// represent different types of Expressions.
+type TreeBuilder interface {
+	// BuildTree will create the tree structure of ExprNodes. The tree structure
+	// of ExprNodes will be traversed in order to build the string representing
+	// different types of Expressions as well as the maps that represent
+	// ExpressionAttributeNames and ExpressionAttributeValues.
+	BuildTree() (ExprNode, error)
+}
+
+// Condition will return the *string corresponding to the Condition Expression
+// of the argument Factory. This method is used to satisfy the members of
+// DynamoDB input structs.
+//
+// Example:
+//
+//     // let factory be an instance of Factory{}
+//
+//     deleteInput := dynamodb.DeleteItemInput{
+//       ConditionExpression:       factory.Condition(),
+//       ExpressionAttributeNames:  factory.Names(),
+//       ExpressionAttributeValues: factory.Values(),
+//       Key: map[string]*dynamodb.AttributeValue{
+//         "PartitionKey": &dynamodb.AttributeValue{
+//           S: aws.String("SomeKey"),
+//         },
+//       },
+//       TableName: aws.String("SomeTable"),
+//     }
+func (factory Factory) Condition() *string {
+	return factory.returnExpression(condition)
+}
+
+// Filter will return the *string corresponding to the Filter Expression of the
+// argument Factory. This method is used to satisfy the members of DynamoDB
+// input structs.
+//
+// Example:
+//
+//     // let factory be an instance of Factory{}
+//
+//     queryInput := dynamodb.QueryInput{
+//       KeyConditionExpression:    factory.KeyCondition(),
+//       FilterExpression:          factory.Filter(),
+//       ExpressionAttributeNames:  factory.Names(),
+//       ExpressionAttributeValues: factory.Values(),
+//       TableName: aws.String("SomeTable"),
+//     }
+func (factory Factory) Filter() *string {
+	return factory.returnExpression(filter)
+}
+
+// Projection will return the *string corresponding to the Projection Expression
+// of the argument Factory. This method is used to satisfy the members of
+// DynamoDB input structs.
+//
+// Example:
+//
+//     // let factory be an instance of Factory{}
+//
+//     queryInput := dynamodb.QueryInput{
+//       KeyConditionExpression:    factory.KeyCondition(),
+//       ProjectionExpression:      factory.Projection(),
+//       ExpressionAttributeNames:  factory.Names(),
+//       ExpressionAttributeValues: factory.Values(),
+//       TableName: aws.String("SomeTable"),
+//     }
+func (factory Factory) Projection() *string {
```

Comment:
Suggest instead of using method for each expression type use a single getter taking the expression type as input.

https://play.golang.org/p/txrOvLujQW